src/backends/gpu/stack.hpp

```diff
@@ -22,91 +25,85 @@ namespace gpu {
 template <typename T>
 class stack {
     using value_type = T;
-    using allocator = memory::managed_allocator<value_type>;
-
-    // The number of items of type value_type that can be stored in the stack
-    unsigned capacity_;
-
-    // The number of items that have been stored
-    unsigned size_;
+    template <typename U>
+    using allocator = memory::managed_allocator<U>;
 
-    // Memory containing the value buffer
-    // Stored in managed memory to facilitate host-side access of values
-    // pushed from kernels on the device.
-    value_type* data_;
+    using base_type = stack_base<value_type>;
+    base_type* base_;
 
 public:
+    stack& operator==(const stack& other) = delete;
+    stack(const stack& other) = delete;
 
-    stack(unsigned capacity):
-        capacity_(capacity), size_(0u)
-    {
-        data_ = allocator().allocate(capacity_);
+    stack(stack&& other) {
+        std::swap(base_, other.base_);
     }
 
-    ~stack() {
-        allocator().deallocate(data_, capacity_);
+    stack& operator=(stack&& other) {
+        std::swap(base_, other.base_);
+        return *this;
+    }
+
+    stack(unsigned capacity) {
+        base_ = allocator<base_type>().allocate(1);
+        base_->capacity = capacity;
+        base_->size = 0;
+        base_->data = allocator<value_type>().allocate(capacity);
     }
 
-    // Append a new value to the stack.
-    // The value will only be appended if do_push is true.
-    __device__
-    void push_back(const value_type& value) {
-        // Atomically increment the size_ counter. The atomicAdd returns
-        // the value of size_ before the increment, which is the location
-        // at which this thread can store value.
-        unsigned position = atomicAdd(&size_, 1u);
-
-        // It is possible that size_>capacity_. In this case, only capacity_
-        // entries are stored, and additional values are lost. The size_
-        // will contain the total number of attempts to push,
-        if (position<capacity_) {
-            data_[position] = value;
+    stack(): stack(0) {}
+
+    ~stack() {
+        if (base_) {
+            allocator<value_type>().deallocate(base_->data, base_->capacity);
+            allocator<base_type>().deallocate(base_, 1);
         }
     }
 
-    __host__
     void clear() {
-        size_ = 0;
+        base_->size = 0;
     }
 
     // The number of items that have been pushed back on the stack.
     // size may exceed capacity, which indicates that the caller attempted
     // to push back more values than there was space to store.
-    __host__ __device__
     unsigned size() const {
-        return size_;
+        return base_->size;
     }
 
```

Comment:
This type is intended for relatively small stacks: it would be performance suicide to use atomic CUDA operations to fill a stack that required something like `2^32` entries.

Reply:
Point taken! I think though a robust solution involving a flag would only have a single write overhead when every thread is attempting to push when the stack is already at capacity, e.g.
```
if (position<s.capacity) {
    s.data[position] = value;
}
else {
    s.overflow_flag = 1; // this flag is initialized to zero on construction, clear()
}
```
What do you think?

Reply:
The `push_back` uses an `atomicAdd()` on `s.size`, and tests the value returned from the atomic operation to decide whether to write. So `size` will get incremented past `capacity` regardless.

Reply:
Oh, I see what you are saying. I go back to my point that it is just unreasonable to use a stack for so many items!